packages/SwingSet/src/liveslots/virtualReferences.js

```diff
@@ -546,16 +559,30 @@ export function makeVirtualReferenceManager(
           doMoreGC = true;
         } else if (recognizer instanceof Set) {
           recognizer.delete(vref);
-        } else if (typeof recognizer === 'function') {
-          recognizer(vref);
+        } else {
+          assert.fail(`unknown recognizer type ${typeof recognizer}`);
         }
       }
     }
+    const prefix = `vom.ir.${vref}|`;
+    let [key] = syscall.vatstoreGetAfter('', prefix);
+    while (key) {
+      syscall.vatstoreDelete(key);
+      const parts = key.split('|');
+      const more = deleteCollectionEntry(parts[1], vref);
+      doMoreGC = doMoreGC || more;
+      [key] = syscall.vatstoreGetAfter(key, prefix);
+    }
     return doMoreGC;
   }
 
   function isVrefRecognizable(vref) {
-    return vrefRecognizers.has(vref);
+    if (vrefRecognizers.has(vref)) {
+      return true;
+    } else {
+      const [key] = syscall.vatstoreGetAfter('', `vom.ir.${vref}|`);
```

Comment:
ooh, this makes me nervous about determinism, we should think it through

`isVrefRecognizable()` might be called by `scanForDeadObjects()` (depending upon `possiblyRetiredSet`, and checks of `getValForSlot`, which is sensitive to GC). It's also called in there for actually-dead vrefs that referred to Presences (imports).

My worry is that doing the vatstore syscall from a GC-sensitive function means we're sometimes making the syscall and sometimes we're not, and that's a recipe for a consensus failure.

I'm also a tiny bit worried that allowing `vrefRecognizers.has` to short-circuit the syscall might make it worse, although I'm not sure `vrefRecognizers.has` is itself GC-sensitive.

Reply:
Oh, I think we're ok here, that's what `dispatch.bringOutYourDead` is for. The transcript will observe a `vatstoreGetAfter` syscall as a consequence of finalizers firing, but only during the consensus delivery of a BOYD. And I think `vrefRecognizers` is only updated by consensus actions.